### ISSUE TITLE
cannot find -lopencv_dep_cudart

### ISSUE THREAD

**Issue by KeeganRen** [Itseez/opencv#6542]
### Please state the information for your system
- OpenCV version: 2.4.12.3
- Host OS: Linux (Ubuntu 14.04)
- Compiler & CMake: GCC 4.9 & CMake 3.5

#####################################################

cannot find -lopencv_dep_cudart

**Comment by KeeganRen** [comment]
When I use OpenCV 2.4.11，it‘s ok，but use OpenCV2.4.12.3，“cannot find -lopencv_dep_cudart”.

include_directories("${PROJECT_SOURCE_DIR}/libclandmark")

find_package( OpenCV REQUIRED)
include_directories(${OpenCV_INCLUDE_DIR})

include_directories(${PROJECT_SOURCE_DIR}/3rd_party/rapidxml-1.13)
include_directories(${PROJECT_SOURCE_DIR}/3rd_party/CImg-1.5.6)

## Several examples using flandmark
set(EXAMPLES static_input video_input)
foreach(var ${EXAMPLES})
    add_executable(${var} ${var}.cpp)
    target_link_libraries(${var} flandmark ${OpenCV_LIBS})
endforeach()

install(TARGETS ${EXAMPLES} EXPORT CLandmarkTargets DESTINATION share/clandmark/examples COMPONENT Examples)

**Comment by KeeganRen** [comment]
## Program
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * Written (W) 2013 Michal Uricar
 * Copyright (C) 2013 Michal Uricar
 */

#include <opencv2/objdetect/objdetect.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>

#include <iostream>
#include <iomanip>
#include <stdio.h>

#include "Flandmark.h"
#include "CFeaturePool.h"
#include "CSparseLBPFeatures.h"
#include "helpers.h"

#define SHOW_WINDOWS

using namespace std;
using namespace cv;
using namespace clandmark;

/** Global variables */
//String face_cascade_name = "lbpcascade_frontalface.xml";
String face_cascade_name = "haarcascade_frontalface_alt.xml";
CascadeClassifier face_cascade;
string window_name = "flandmark - static input demo";
RNG rng(12345);

cimg_library::CImg<unsigned char> * cvImgToCImg(cv::Mat &cvImg)
{
	cimg_library::CImg<unsigned char> * result = new cimg_library::CImg<unsigned char>(cvImg.cols, cvImg.rows);

	for (int x = 0; x < cvImg.cols; ++x)
		for (int y = 0; y < cvImg.rows; ++y)
			(*result)(x, y) = cvImg.at<uchar>(y, x);

	return result;
}

cv::Mat & CImgtoCvImg(cv::Mat &result, cimg_library::CImg<unsigned char> *img)
{
	result = cv::Mat(img->height(), img->width(), CV_8U);

	for (int x=0; x < result.cols; ++x)
		for (int y=0; y < result.rows; ++y)
			result.at<uchar>(y, x) = (*img)(x, y);

	return result;
}

/** @function detectAndDisplay */
void detectAndDisplay( Mat &frame, Flandmark *flandmark, CFeaturePool *featurePool)
{
	std::vector<Rect> faces;
	Mat frame_gray;
	int bbox[8];
	fl_double_t *landmarks;

	cvtColor( frame, frame_gray, CV_BGR2GRAY );
//    cvtColor( frame, frame_gray, COLOR_BGR2GRAY );  // <- OpenCV 3.0
	//equalizeHist( frame_gray, frame_gray );

	//-- Detect faces
	face_cascade.detectMultiScale( frame_gray, faces, 1.1, 2, 0|CV_HAAR_SCALE_IMAGE, Size(30, 30) );
//    face_cascade.detectMultiScale( frame_gray, faces, 1.1, 2, 0|CASCADE_SCALE_IMAGE, Size(30, 30) );

	for( uint32_t i = 0; i < faces.size(); i++ )
	{
		// Get detected face bounding box
		bbox[0] = faces[i].x;
		bbox[1] = faces[i].y;
		bbox[2] = faces[i].x+faces[i].width;
		bbox[3] = faces[i].y;
		bbox[4] = faces[i].x+faces[i].width;
		bbox[5] = faces[i].y+faces[i].height;
		bbox[6] = faces[i].x;
		bbox[7] = faces[i].y+faces[i].height;

		// Detect facial landmarks
		cimg_library::CImg<unsigned char>* frm_gray = cvImgToCImg(frame_gray);
//		flandmark->detect_from_nf(frm_gray, bbox);
//		flandmark->detect( frm_gray, bbox );
		flandmark->detect_optimized(frm_gray, bbox);

		delete frm_gray;

		// Get detected landmarks
		landmarks = flandmark->getLandmarks();

		// Draw bounding box and detected landmarks
//		rectangle(frame, Point(bbox[0], bbox[1]), Point(bbox[2], bbox[3]), Scalar(255, 0, 0));
		circle(frame, Point(int(landmarks[0]), int(landmarks[1])), 2, Scalar(255, 0, 0), -1);
		for (int i=2; i < 2*flandmark->getLandmarksCount(); i+=2)
		{
			circle(frame, Point(int(landmarks[i]), int(landmarks[i+1])), 2, Scalar(0, 0, 255), -1);
		}

		// Textual output
		printTimingStats(flandmark->timings);
		printLandmarks(landmarks, flandmark->getLandmarksCount());
		printLandmarks(flandmark->getLandmarksNF(), flandmark->getLandmarksCount());
	}

	//-- Show what you got
#ifdef SHOW_WINDOWS
	imshow( window_name, frame );
#endif
}

/** @function main */
int main( int argc, const char** argv )
{
	if (argc < 4)
	{
		cerr << "Usage: static_input <path_to_haarcascade> <flandmark_model.xml> <input_image> [<output_image>]" << endl;
		return -1;
	}

	double tim;
	Mat image;

	//Flandmark *flandmark = new Flandmark();

	tim = (double)getTickCount();

	Flandmark *flandmark = Flandmark::getInstanceOf(argv[2]);
	if (!flandmark)
	{
		cerr << "Usage: static_input <flandmark_model.xml> <input_image> [<output_image>]" << endl;
		return -1;
	}

	CFeaturePool *featurePool = new CFeaturePool(flandmark->getBaseWindowSize()[0], flandmark->getBaseWindowSize()[1]);
	featurePool->addFeaturesToPool(
				new CSparseLBPFeatures(
					featurePool->getWidth(),
					featurePool->getHeight(),
					featurePool->getPyramidLevels(),
					featurePool->getCumulativeWidths()
					)
				);

	flandmark->setNFfeaturesPool(featurePool);

	tim = ((double)getTickCount() - tim)/getTickFrequency() * 1000;

	cout << "Flandmark model loaded in " << tim << " ms" << endl;

	//-- 1. Load the cascades
	if( !face_cascade.load( argv[1]+face_cascade_name ) )
	{
		printf("--(!)Error loading\n");
		return -1;
	};

#ifdef SHOW_WINDOWS
	namedWindow(window_name, CV_WINDOW_KEEPRATIO);
//  //namedWindow(window_name, WINDOW_KEEPRATIO);
#endif

	// Read input image
	image = imread(argv[3]);

	if (!image.empty())
	{
		detectAndDisplay( image, flandmark, featurePool );
	} else {
		cout << "Wrong input." << endl;
	}

	if (argc > 4)
	{
		imwrite(argv[4], image);
	}

#ifdef SHOW_WINDOWS
	waitKey();
#endif

	delete featurePool;
	delete flandmark;

	return 0;
}

**Comment by jet47** [comment]
Try to set `CUDA_USE_STATIC_CUDA_RUNTIME` CMake variable to `OFF`.

**Comment by KeeganRen** [comment]
@jet47 Thanks!

**Comment by StevenPuttemans** [comment]
I guess it is solved then!

**Comment by Deepak-** [comment]
@KeeganRen I have the same error. Did you just add it to cmake like this `-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF`

**Comment by romanlarionov** [comment]
I have this same issue. I'm running a clean install of Ubuntu 14.04 with OpenCV version 2.4.13.

After starting a brand new project following [these](http://docs.opencv.org/2.4/doc/tutorials/introduction/linux_gcc_cmake/linux_gcc_cmake.html) exact instructions, I receive the same error.

**Comment by mlamarre** [comment]
I also had the same error and so as this guy:
http://stackoverflow.com/questions/37534604/opencv-with-cmake-version-3-5-2-vs-cmake-2-x-x

The above solution is only a workaround. Not setting this variable leads to a wrong cmake script to be installed in /usr/local/share/OpenCV/OpenCVConfig.cmake.

This script fails to set IMPORTED_LOCATION for libcudart.so because the CUDA_LIBRARIES list contains the path to libcudart_static.a instead of the libcudart.so. However, one of the main .so file of OpenCV still depends on this shared lib, thus the error "cannot find -lopencv_dep_cudart"

I got this error while building Caffe on 16.04. 

Maybe CUDA_USE_STATIC_CUDA_RUNTIME=OFF should be the default.

**Comment by abrock** [comment]
I also had this problem on debian testing 64 bit, opencv-2.4.13.zip. I couldn't solve it by setting `-D CUDA_USE_STATIC_CUDA_RUNTIME=OFF` so I removed all instances of the string "opencv_dep_cudart;" from the installed .cmake files (dirty hack).

**Comment by thalin** [comment]
I'm trying to build OpenCV 2.4 with CUDA 8 support and I'm also getting this issue. Building with `-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF` doesn't prevent `/usr/share/OpenCV/OpenCVModules-release.cmake` from having `IMPORTED_LINK_INTERFACE_LIBRARIES_RELEASE "opencv_dep_cudart;opencv_dep_nppc;opencv_dep_nppi;opencv_dep_npps` populated. I don't think that this is a reasonable workaround.

**Comment by jet47** [comment]
Did you build your own code with `-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF`?

**Comment by zhanghang1989** [comment]
``cmake -D CUDA_USE_STATIC_CUDA_RUNTIME=OFF `` fixed this problem for me. Thank you guys!

**Comment by rain16881** [comment]
in the CmakeLists.txt
set(CUDA_USE_STATIC_CUDA_RUNTIME OFF)

In the ide clion

**Comment by rain16881** [comment]
but have new error 
OpenCV Error: Gpu API call (invalid configuration argument) in call, file XXX/opencv-2.4.13/modules/gpu/include/opencv2/gpu/device/detail/transform_detail.hpp, line 361
Error: XXX/opencv-2.4.13/modules/gpu/include/opencv2/gpu/device/detail/transform_detail.hpp:361: error: (-217) invalid configuration argument in function call

**Comment by metorm** [comment]
`-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF` fixes my problem on Jetson TX-1, ubuntu 16.04, cuda 8

**Comment by antran89** [comment]
@rain16881 Your solution does not work for me. But `-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF` works!

**Comment by tneumann** [comment]
I also had this issue on Windows and was not able to solve this, since FindCUDA wouldn't let me turn off CUDA_USE_STATIC_CUDA_RUNTIME (I'm not sure why). I found another workaround by simply producing a "fake" opencv_dep_cudart.lib. This works since OpenCV actually does not seem to require any code from this .lib. 

In your CMakeLists.txt that references OpenCV, simply add:
```
ADD_LIBRARY(opencv_dep_cudart STATIC <an_empty_file.cpp>)
TARGET_LINK_LIBRARIES(your_app opencv_dep_cudart)
```

Again, this is just a dirty workaround that worked for my case on Windows with Visual Studio 2015.

**Comment by santonatos** [comment]
-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF solves my problem on Jetson TX-1..very helpful thanks

**Comment by jparrish88** [comment]
Since it wasn't mentioned, -DCUDA_USE_STATIC_CUDA_RUNTIME=OFF solves my problem on a Jetson TX2

**Comment by crazysnowboy** [comment]
Using option(CUDA_USE_STATIC_CUDA_RUNTIME OFF) on CMakeLists.txt file has fixed my problem.

**Comment by SkyCrawl** [comment]
In my case, I only managed to resolve the issue by removing all instances of `-lopencv_dep_cudart` from the `*.link.txt` files generated by `cmake` command, prior to executing `make` and `make install`.

Cd into OpenALPR's root folder and use `grep -rl "opencv_dep_cudart" .` to locate all the affected files.

**Comment by sputnick1124** [comment]
For anyone having this issue trying to build a ROS package with catkin-tools, setting the option with `catkin config --cmake-args -D(CUDA_USE_STATIC_CUDA_RUNTIME=OFF` and then `catkin build` worked for me. This is on a TX1 with Cuda 8 and Ubuntu 16.04.

**Comment by wscnwps** [comment]
Make a soft link in /usr/lib to "/usr/local/cuda/lib64/libcudart.so". E.g. 
    `sudo ln -s /usr/local/cuda/lib64/libcudart.so /usr/lib/libopencv_dep_cudart.so`

**Comment by jackyko1991** [comment]
i think this bug need to be fixed...some packages like caffe, SimpleITK cannot turn `CUDA_USE_STATIC_CUDA_RUNTIME=OFF`. Each time I switch this option off it will automatically change back to `ON`

**Comment by Haffon** [comment]
I have the same problem when compile some dense optical flow code with cuda 8.0 and opencv 2.4.13.5(which compiled with CUDA enabled), and I fixed it by
cmake -D CUDA_USE_STATIC_CUDA_RUNTIME=OFF .
thanks your guys.

**Comment by northeastsquare** [comment]
> `-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF` fixes my problem on Jetson TX-1, ubuntu 16.04, cuda 8
> 
> update at 2016-9-18 13:30
> 
> Well, it seems compiling with `-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF` results in the follwoing link issue:
> 
> `opencv_dep_cudart;opencv_dep_nppc;opencv_dep_nppi;opencv_dep_npps`
> 
> @thalin You are right. So, what is your solution fianlly?
> 
> update at 2016-9-18 13:59
> 
> OK, I fixed this. I got the `opencv_dep_cudart;opencv_dep_nppc;opencv_dep_nppi;opencv_dep_npps` issue because I missed
> 
> ```
> find_package( OpenCV REQUIRED)
> 
> ......
> 
> target_link_libraries(${var}  ${OpenCV_LIBS})
> ```

是在cmake/OpenCVModule.cmake这里定义的： ocv_add_module(yaom INTERNAL opencv_core opencv_highgui opencv_flann OPTIONAL opencv_gpu)macro(ocv_add_module _name) string(TOLOWER "${_name}" name) set(the_module opencv_${name})。opencv npp的库都加上了opencv_这个前缀

**Comment by zhaoxuyan** [comment]
> `-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF` fixes my problem on Jetson TX-1, ubuntu 16.04, cuda 8
> 
> update at 2016-9-18 13:30
> 
> Well, it seems compiling with `-DCUDA_USE_STATIC_CUDA_RUNTIME=OFF` results in the follwoing link issue:
> 
> `opencv_dep_cudart;opencv_dep_nppc;opencv_dep_nppi;opencv_dep_npps`
> 
> @thalin You are right. So, what is your solution fianlly?
> 
> update at 2016-9-18 13:59
> 
> OK, I fixed this. I got the `opencv_dep_cudart;opencv_dep_nppc;opencv_dep_nppi;opencv_dep_npps` issue because I missed
> 
> ```
> find_package( OpenCV REQUIRED)
> 
> ......
> 
> target_link_libraries(${var}  ${OpenCV_LIBS})
> ```

我解决了这个问题~~
```
ln -s /usr/local/cuda/lib64/libcudart.so /usr/lib/libopencv_dep_cudart.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppial.so /usr/local/lib/libopencv_dep_nppial.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppicc.so /usr/local/lib/libopencv_dep_nppicc.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppicom.so /usr/local/lib/libopencv_dep_nppicom.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppidei.so /usr/local/lib/libopencv_dep_nppidei.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppif.so /usr/local/lib/libopencv_dep_nppif.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppig.so /usr/local/lib/libopencv_dep_nppig.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppim.so /usr/local/lib/libopencv_dep_nppim.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppist.so /usr/local/lib/libopencv_dep_nppist.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppisu.so /usr/local/lib/libopencv_dep_nppisu.so
ln -s /usr/local/cuda-9.0/targets/x86_64-linux/lib/libnppitc.so /usr/local/lib/libopencv_dep_nppitc.so
```
